### PR TITLE
Box symbol should list pins in the order they appear in the pin-properties table.

### PR DESCRIPTION
**Pull request by CuppoJava**
Sort all pins by the order in which they are referred to in the pin-properties table before creating the box symbol.

### CODE REVIEW

**Review by ariel-hi — DISMISSED** [review]
Reviewing in tandem with [JITPCB PR 53](https://github.com/JITx-Inc/jitpcb/pull/536).

When a box symbol is made, pins are sorted inside the pcb-component according to their `pin-properties-row-index` property. Specialized comparator.

Nice.

File: utils/box-symbol.stanza [hunk 1]
```diff
@@ -244,6 +244,22 @@ public defn make-box-symbol (banks:Seqable<BoxBank>) :
 ;=======================================================<doc>
 
 public defn make-box-symbol () :
+  ;Sort the given pins by their .pin-properties-row-index field.
+  defn sort-by-row-index (pins:Tuple<JITXObject>) -> Tuple<JITXObject> :
+    ;Compare pin[i0] versus pin[i1].
+    ;Indices are used, instead of the pin directly, in order to perform a
+    ;stable sort.
+    ;All pins without a .pin-properties-row-index is put in the beginning.
```
Comment (bpellenard) [inline utils/box-symbol.stanza:252]: are put

File: utils/box-symbol.stanza [hunk 2]
```diff
@@ -244,6 +244,22 @@ public defn make-box-symbol (banks:Seqable<BoxBank>) :
...
+    defn compare-index (i0:Int, i1:Int) :
+      val a = pins[i0]
+      val b = pins[i1]
+      match(value?(property?(a.pin-properties-row-index)), value?(property?(b.pin-properties-row-index))) :
+        (ai:Int, bi:Int) : compare(ai, bi)
+        (ai:Int, bi) : 1
+        (ai, bi:Int) : -1
+        (ai, bi) : compare(i0,i1)
+    val indices = qsort(0 to length(pins), compare-index)
```
Comment (bpellenard) [inline utils/box-symbol.stanza:261]: `qsort` or `qsort!`

File: utils/box-symbol.stanza [hunk 3]
```diff
@@ -244,6 +244,22 @@ public defn make-box-symbol (banks:Seqable<BoxBank>) :
...
 
 public defn make-box-symbol () :
+  ;Sort the given pins by their .pin-properties-row-index field.
+  defn sort-by-row-index (pins:Tuple<JITXObject>) -> Tuple<JITXObject> :
+    ;Compare pin[i0] versus pin[i1].
+    ;Indices are used, instead of the pin directly, in order to perform a
+    ;stable sort.
+    ;All pins without a .pin-properties-row-index is put in the beginning.
+    defn compare-index (i0:Int, i1:Int) :
+      val a = pins[i0]
```
Comment (bpellenard) [inline utils/box-symbol.stanza:254]: `val a = value?(property?(pins[i0].pin-properties-row-index))`

File: utils/box-symbol.stanza (hunk 2 above)
Comment (ariel-hi) [reply utils/box-symbol.stanza:261]: `qsort!` modifies the passed-in collection.
`qsort` returns a new collection.

File: utils/box-symbol.stanza [hunk 4]
```diff
@@ -244,6 +244,22 @@ public defn make-box-symbol (banks:Seqable<BoxBank>) :
...
+      val b = pins[i1]
+      match(value?(property?(a.pin-properties-row-index)), value?(property?(b.pin-properties-row-index))) :
```
Comment (bpellenard) [inline utils/box-symbol.stanza:256]:
```
val a = value?(property?(pins[i0].pin-properties-row-index))
val b = value?(property?(pins[i1].pin-properties-row-index))`
match(a, b) :
```

File: utils/box-symbol.stanza (hunk 1 above)
Comment (CuppoJava) [reply utils/box-symbol.stanza:252]: Thanks!

File: utils/box-symbol.stanza (hunk 4 above)
Comment (CuppoJava) [reply utils/box-symbol.stanza:256]: Thanks. That's cleaner.

**Review by bpellenard — APPROVED** [review]

File: utils/box-symbol.stanza [hunk 5]
```diff
@@ -244,6 +244,24 @@ public defn make-box-symbol (banks:Seqable<BoxBank>) :
...
 public defn make-box-symbol () :
+  ;Sort the given pins by their .pin-properties-row-index field.
+  defn sort-by-row-index (pins:Tuple<JITXObject>) -> Tuple<JITXObject> :
+    ;Compare pin[i0] versus pin[i1].
+    ;Indices are used, instead of the pin directly, in order to perform a
+    ;stable sort.
+    ;All pins without a .pin-properties-row-index are put in the beginning.
+    defn compare-index (i0:Int, i1:Int) :
+      val a = pins[i0]
+      val b = pins[i1]
```
Comment (bpellenard) [inline utils/box-symbol.stanza:255]: minor but you does not really need these two lines

Comment (CuppoJava) [reply utils/box-symbol.stanza]: You're right. I would like it to be inlined too.

**Review by bpellenard — APPROVED** [review]

**Pull request merged**